drivers/rtc/rtc_rv8803.c

```diff
@@ -0,0 +1,861 @@
+#define RV8803_CONTROL_AIE_BIT   BIT(3)
+#define RV8803_CONTROL_EIE_BIT   BIT(2)
+#define RV8803_CONTROL_RESET_BIT BIT(0)
+
+#define RV8803_MONDAY_MASK    BIT(0)
+#define RV8803_TUESDAY_MASK   BIT(1)
+#define RV8803_WEDNESDAY_MASK BIT(2)
+#define RV8803_THURSDAY_MASK  BIT(3)
+#define RV8803_FRIDAY_MASK    BIT(4)
+#define RV8803_SATURDAY_MASK  BIT(5)
+#define RV8803_SUNDAY_MASK    BIT(6)
+
+#define RV8803_OFFSET_MASK GENMASK(5, 0)
+
+/* Offset between first tm_year and first RV8803 year */
+#define RV8803_YEAR_OFFSET (2000 - 1900)
+
+/* RV8803 enumerates months 1 to 12 */
+#define RV8803_MONTH_OFFSET -1
+
+/* Max value of seconds, needed for readout procedure workaround */
+#define RV8803_SECONDS_MAX_VALUE 59
+
+/* See RV-8803-C7 Application Manual p. 22, 3.9. */
+#define RV8803_OFFSET_PPB_PER_LSB    238
+#define RV8803_OFFSET_PPB_MIN        (-32 * RV8803_OFFSET_PPB_PER_LSB)
+#define RV8803_OFFSET_PPB_MAX        (31 * RV8803_OFFSET_PPB_PER_LSB)
+#define RV8803_OFFSET_SIGN_BIT_INDEX 5 /* Required for aging offset sign extension */
+
+/* CLKOUT property enum values */
+#define RV8803_PROP_ENUM_1HZ     0
+#define RV8803_PROP_ENUM_1024HZ  1
+#define RV8803_PROP_ENUM_32768HZ 2
+
+#define DT_DRV_COMPAT microcrystal_rv8803
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "Micro Crystal RV8803 driver enabled without any devices"
+#endif
+
+/* RTC time fields supported by RV8803 */
+#define RV8803_RTC_TIME_MASK                                                                       \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_YEAR |     \
+	 RTC_ALARM_TIME_MASK_WEEKDAY)
+
+/* RTC alarm time fields supported by RV8803 */
+#define RV8803_RTC_ALARM_TIME_MASK                                                                 \
+	(RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MONTHDAY |    \
+	 RTC_ALARM_TIME_MASK_WEEKDAY)
+
+/* Helper macro to guard GPIO interrupt related stuff */
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios) &&                                                 \
+	(defined(CONFIG_RTC_ALARM) || defined(CONFIG_RTC_UPDATE))
+#define RV8803_INT_GPIOS_IN_USE 1
+#endif
+
+LOG_MODULE_REGISTER(rv8803, CONFIG_RTC_LOG_LEVEL);
+
+struct rv8803_config {
+	const struct i2c_dt_spec i2c;
+#ifdef RV8803_INT_GPIOS_IN_USE
+	struct gpio_dt_spec gpio_int;
+#endif
+	uint16_t clkout_freq;
+};
+
+struct rv8803_data {
+	struct k_sem lock;
+#ifdef RV8803_INT_GPIOS_IN_USE
+	const struct device *dev;
+	struct gpio_callback irq_callback;
+	struct k_work work;
+
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+#endif
+
+#ifdef CONFIG_RTC_UPDATE
+	rtc_update_callback update_callback;
+	void *update_user_data;
+#endif
+#endif
+};
+
+static void rv8803_lock_sem(const struct device *dev)
+{
+	k_sem_take(&((struct rv8803_data *)dev->data)->lock, K_FOREVER);
```

Comment:
Please declare and typecast the data before accessing it, both for consistency and clarity. Its even free :)
```
const struct rv8803_data *data = dev->data; /* no explicit typecast required */

k_sem_take(&data->data)->lock, K_FOREVER);
```

Reply:
You're right, fixed!